File: Assets/GeometryAlgorithm/SolidSpanGroup.cs
using Mathd;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Geometry_Algorithm
{
    /// <summary>
    /// 实心跨距
    /// </summary>
    public struct SolidSpan
    {
        public double startPos;
        public double endPos;

        public int startCellIdx;
        public int endCellIdx;
    }

    
    /// <summary>
    /// 实心跨距组
    /// </summary>
    public class SolidSpanGroup
    {
        public Dictionary<int, LinkedList<SolidSpan>> soildSpanDict = new Dictionary<int, LinkedList<SolidSpan>>();

        public void AppendVoxBox(VoxBox voxBox)
        {
            int key = GetKey(voxBox.floorCellIdxX, voxBox.floorCellIdxZ);
            LinkedList<SolidSpan> cellSpanList;

            if (soildSpanDict.TryGetValue(key, out cellSpanList) == false)
            {
                cellSpanList = new LinkedList<SolidSpan>();
                soildSpanDict[key] = cellSpanList;
            }

            AppendVoxBoxToSpanHeightList(cellSpanList, voxBox);
        }

        void AppendVoxBoxToSpanHeightList(LinkedList<SolidSpan> cellSpanList, VoxBox voxBox)
        {
            int voxStartIdx = voxBox.heightCellStartIdx;
            int voxEndIdx = voxBox.heightCellStartIdx;
            double yPosStart = voxBox.yPosRange[0];
            double yPosEnd = voxBox.yPosRange[1];    

            LinkedListNode<SolidSpan> startNode = null;
            LinkedListNode<SolidSpan> endNode = null;

            var node = cellSpanList.First;
            for (; node != null; node = node.Next)
            {
                if(node.Value.startCellIdx > voxStartIdx && startNode == null)
                {
                    startNode = node;
                }
                else if (voxStartIdx >= node.Value.startCellIdx  &&
                    voxStartIdx <= node.Value.endCellIdx)
                {
                    yPosStart = node.Value.startPos;
                    voxStartIdx = node.Value.startCellIdx;
                    startNode = node;
                }

                if (node.Value.startCellIdx > voxEndIdx && endNode == null)
                {
                    endNode = node.Previous;
                    break;
                }
                else if (voxEndIdx >= node.Value.startCellIdx && 
                    voxEndIdx <= node.Value.endCellIdx)
                {
                    yPosEnd = node.Value.endPos;
                    voxEndIdx = node.Value.endCellIdx;
                    endNode = node;
                    break;
                }
            }

            if(startNode != null && endNode == null)
                endNode = cellSpanList.Last;

            SolidSpan voxSpan = new SolidSpan()
            {
                startPos = yPosStart,
                endPos = yPosEnd,
                startCellIdx = voxStartIdx,
                endCellIdx = voxEndIdx
            };

            if (startNode == null && endNode == null)
            {
                if(node == cellSpanList.First)
                    cellSpanList.AddFirst(voxSpan);
                else
                    cellSpanList.AddLast(voxSpan);
            }     
            else
            {
                var prevNode = startNode.Previous;
                var mnode = startNode;
                LinkedListNode<SolidSpan> tmpNode;
                bool flag = true;

                while(flag)
                {
                    if (mnode == endNode)
                        flag = false;

                    tmpNode = mnode.Next;
                    cellSpanList.Remove(mnode);
                    mnode = tmpNode;
                }

                if(prevNode == null)
                    cellSpanList.AddFirst(voxSpan);
                else
                    cellSpanList.AddAfter(prevNode, voxSpan);
            }
        }

        public int GetKey(int cellx, int cellz)
        {
            if (cellx < 0 || cellz < 0)
                return -1;

            return (cellx << 14) | cellz; 
        }

        public int[] GetCellIdxs(int key)
        {
            int[] idxs = new int[] { key >> 14, 0x3FFF & key };
            return idxs;
        }

    }
}

File: MiNavMeshPlus/core/SolidSpanGroup.cpp
#include"SolidSpanGroup.h"

SolidSpanGroup::SolidSpanGroup(VoxelSpace* voxSpace)
{
	this->voxSpace = voxSpace;
	solidSpanGrids = voxSpace->solidSpanGrids;
	gridCount = voxSpace->gridCount;
}

SolidSpanGroup::~SolidSpanGroup()
{

}

//test
void SolidSpanGroup::AppendVoxBox(
	int floorCellIdxX, int floorCellIdxZ,
	int heightCellStartIdx, int heightCellEndIdx)
{
	int idx = voxSpace->GetFloorGridIdx(floorCellIdxX, floorCellIdxZ);
	SolidSpanList* solidSpanList = &(solidSpanGrids[idx]);

	SolidSpanList tmp = *solidSpanList;

	if (solidSpanList->first == nullptr)
	{
		solidSpanList->floorCellIdxX = floorCellIdxX;
		solidSpanList->floorCellIdxZ = floorCellIdxZ;
	}

	AppendVoxBoxToSpanHeightList(solidSpanList, heightCellStartIdx, heightCellEndIdx);
}


void SolidSpanGroup::AppendVoxBoxToSpanHeightList(
	SolidSpanList* solidSpanList,
	int heightCellStartIdx, int heightCellEndIdx)
{
	int voxStartIdx = heightCellStartIdx;
	int voxEndIdx = heightCellEndIdx;
	float yPosStart = voxStartIdx * voxSpace->cellHeight;
	float yPosEnd = voxEndIdx * voxSpace->cellHeight;

	SolidSpan* startNode = nullptr;
	SolidSpan* endNode = nullptr;

	SolidSpan* node = solidSpanList->first;
	for (; node != nullptr; node = node->next)
	{
		if (startNode == nullptr)
		{
			if (node->ystartCellIdx > voxStartIdx)
			{
				startNode = node;
			}
			else if (voxStartIdx >= node->ystartCellIdx &&
				voxStartIdx <= node->yendCellIdx)
			{
				yPosStart = node->ystartPos;
				voxStartIdx = node->ystartCellIdx;
				startNode = node;
			}
		}

		if (endNode == nullptr)
		{
			if (node->ystartCellIdx > voxEndIdx)
			{
				endNode = node->prev;
				break;
			}
			else if (voxEndIdx >= node->ystartCellIdx &&
				voxEndIdx <= node->yendCellIdx)
			{
				yPosEnd = node->yendPos;
				voxEndIdx = node->yendCellIdx;
				endNode = node;
				break;
			}
		}
	}

	if (startNode != nullptr && node == nullptr)
		endNode = solidSpanList->last;


	SolidSpan* voxSpan = voxSpace->GetSoildSpan();
	voxSpan->ystartPos = yPosStart;
	voxSpan->yendPos = yPosEnd;
	voxSpan->ystartCellIdx = voxStartIdx;
	voxSpan->yendCellIdx = voxEndIdx;

	if (endNode != nullptr && endNode->next == startNode)
	{
		solidSpanList->AddAfter(endNode, voxSpan);
		return;
	}

	if (startNode != nullptr && endNode == nullptr)
	{
		solidSpanList->AddFirst(voxSpan);
		return;
	}

	if (startNode == nullptr && endNode == nullptr)
	{
		if (node == solidSpanList->first)
			solidSpanList->AddFirst(voxSpan);
		else
			solidSpanList->AddLast(voxSpan);
	}
	else
	{
		SolidSpan* prevNode = startNode->prev;
		SolidSpan* mnode = startNode;
		SolidSpan* tmpNode;
		bool flag = true;

		while (flag)
		{
			if (mnode == endNode)
				flag = false;

			tmpNode = mnode->next;
			solidSpanList->Remove(mnode);
			mnode = tmpNode;
		}

		if (prevNode == nullptr)
			solidSpanList->AddFirst(voxSpan);
		else
			solidSpanList->AddAfter(prevNode, voxSpan);
	}
}